### PR TITLE
68 20230829 replication referential integrity

### PR DESCRIPTION
**Pull request by Firstyear**
Relates #68 - implements referential integrity over replication. This adds tests for some of the super gnarly edge cases and ensures that everything is all consistent and nice. 

Checklist

- [ x ] This pr contains no AI generated code
- [ x ] cargo fmt has been run
- [ x ] cargo clippy has been run
- [ x ] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

File: server/lib/src/plugins/refint.rs [hunk 1]
```diff
@@ -242,9 +364,10 @@ impl ReferentialIntegrity {
                 c.attribute_equality(Attribute::Class.as_ref(), &EntryClass::DynGroup.into());
 
             ref_types.values().filter_map(move |rtype| {
-                // Skip dynamic members
+                // Skip dynamic members, these are recalculated by the
+                // memberof plugin.
                 let skip_mb = dyn_group && rtype.name == "dynmember";
-                // Skip memberOf.
+                // Skip memberOf, also recalculated.
                 let skip_mo = rtype.name == "memberof";
```
Comment (yaleman) [inline server/lib/src/plugins/refint.rs:371]: Since we're messing around in this part of the code, can you please use the static objects I added the other day instead of strings for memberof/dynmember?

**Review by yaleman — APPROVED** [review]

File: server/lib/src/plugins/refint.rs [hunk 2]
```diff
@@ -50,16 +50,91 @@ impl ReferentialIntegrity {
...
     }
+
+    fn check_uuids_exist_slow(
+        qs: &mut QueryServerWriteTransaction,
+        inner: &[Uuid],
+    ) -> Result<Vec<Uuid>, OperationError> {
+        if inner.is_empty() {
+            // There is nothing to check! Move on.
+            // Should be unreachable.
+            trace!("no reference types modified, skipping check");
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply server/lib/src/plugins/refint.rs:66]: I think we still need to handle the empty case though, because it's valid for a mod to change no references.

File: server/lib/src/plugins/refint.rs (hunk 1 above)
Comment (Firstyear) [reply server/lib/src/plugins/refint.rs:371]: Opps mybad I'll fix that up next time :)

File: server/lib/src/plugins/refint.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (yaleman) [reply server/lib/src/plugins/refint.rs:66]: ah ok, just the "should be unreachable" bit was confusing :smile:
